### PR TITLE
chore: refactor BlockClock contract & create ERC20 TimeToken

### PR DESCRIPTION
**Pull request by mevtxn**

### CODE REVIEW

File: src/utilities/BlockTime.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    uint256 public currentEarthTimeAvg;
+
+    event Tick(uint256 currentEarthTimeBlockStart, uint256 currentEarthTimeBlockEnd);
+    event EarthTimeUpdated(uint256 newEarthTime, Chronicle[] chronicles);
+
+    error InsufficientEvidence();
+
+    constructor() {
+        _grantRole(DEFAULT_ADMIN_ROLE, _msgSender());
+        _grantRole(ADMIN_ROLE, _msgSender());
```
[the remark this answers is not in the document]
Comment (mevtxn) [reply src/utilities/BlockTime.sol:37]: Using the SCHEDULER_ROLE. After deployment of BlockTime & BlockScheduler, the admin will call the grantRole external function to provide the scheduler role to the _blockScheduler contract. 
`_grantRole(SCHEDULER_ROLE, _blockScheduler);
`

File: test/examples/BlockTime/BlockTimeScheduler.sol [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            addr: address(this),
+            gas: 1000000,
+            callvalue: abi.encodeWithSignature("verifySignature(bytes)", data)
+        });
+
+        assertFutureCallTo(callObj, 1);
+    }
+
+    /// @dev The onlyOwner modifier will be later changed to execute calls through a governance proposal
+    function  batchMint()  external onlyOwner {
```
Comment (TokenTitan) [inline test/examples/BlockTime/BlockTimeScheduler.sol:64]: Should be internal and executed from within the updateTime function

File: test/examples/BlockTime/BlockTimeScheduler.sol [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    /// @notice function to be checked by Laminator before rescheduling a call to disburseKITNs
+    /// @dev The onlyOwner modifier will be later changed to execute calls through a governance proposal
+    function setContinue(bool _shouldContinue) external onlyOwner {
+        shouldContinue = _shouldContinue;
+    }
+
+    /// @notice function to be called by solver to ensure a succesful and valid call
+    function verifySignature(bytes calldata /* data */ ) public view {
```
Comment (TokenTitan) [inline test/examples/BlockTime/BlockTimeScheduler.sol:88]: Is not needed here

**Review by TokenTitan — APPROVED** [review]

File: test/examples/BlockTime/BlockTimeScheduler.sol (hunk 3 above)
Comment (mevtxn) [reply test/examples/BlockTime/BlockTimeScheduler.sol]: Removing this

File: test/examples/BlockTime/BlockTimeScheduler.sol [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import "openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
+
+interface IBlockTime {
+    function moveTime(BlockTime.Chronicle[] calldata chronicles, uint256 meanCurrentEarthTime) external;
+}
+
+/**
+ * @notice This is an POC example of a block scheduler
+ */
+contract BlockTimeScheduler is SmarterContract, Ownable {
```
[the remark this answers is not in the document]
Comment (mevtxn) [reply test/examples/BlockTime/BlockTimeScheduler.sol:13]: Will create a different branch for that. You can deploy it parallelly after merging this PR

File: src/utilities/BlockTime.sol [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        timeToken = new TimeToken(_name, _symbol, address(this));
+        _grantRole(DEFAULT_ADMIN_ROLE, _msgSender());
+        _grantRole(ADMIN_ROLE, _msgSender());
+    }
+
+    /// @notice changes earth avg time
+    function moveTime(Chronicle[] calldata chronicles, uint256 meanCurrentEarthTime, address[] calldata receivers, uint256[] calldata amounts) external onlyRole(SCHEDULER_ROLE) nonReentrant {
+        currentEarthTimeAvg = meanCurrentEarthTime;
+        timeToken.batchMint(receivers, amounts);
+        emit EarthTimeUpdated(meanCurrentEarthTime, chronicles);
```
Comment (TokenTitan) [inline src/utilities/BlockTime.sol:45]: Should also emit receivers and amounts

Comment (mevtxn) [reply src/utilities/BlockTime.sol:45]: It is being emitted in the TimeToken contract

Comment (TokenTitan) [reply src/utilities/BlockTime.sol:45]: Please add it here as well

**Review by TokenTitan — APPROVED** [review]

**Pull request merged**